### DISCUSSION TITLE
Como vocês desenvolvem Backend com JavaScript hoje?

### DISCUSSION THREAD

**Discussion by Cleysonlb** [frontendbr/forum#585] [Perguntas e Respostas]
Fala galera beleza? estou na duvida de que ferramentas usar no backend com javascript (NODE), existe muitas opções e nada muito "padrão" tipo, que ORM (Object-relational mapping) vocês usam?
Costumam usar banco relacional (postgresql) ou banco não relacional?
Sei que muito das especificações dependem do projeto, mas estou buscando em um contexto geral mesmo para conseguir um norte! :1st_place_medal:

**Comment by rafael1995** [comment]
Fala @Cleysonlb , blz? 
Não sei se entendi muito bem a pergunta, mas não é muito recomendado utilizar o Javascript no back-end de uma aplicação, por questões de segurança, ele roda no lado cliente. 
P/ isso recomendo dar uma pesquisada  a respeito do noje.js

**Comment by Cleysonlb** [comment]
Opa @rafael1995 é justamente sobre node que to falando hahah, Junto com ele qual o "kit" de ferramentas ta sendo legal usar pra usar junto?

**Comment by devigor** [comment]
O que eu vejo mais o pessoal usando é o Node com MongoDB, em alguns casos já vi usando também com o MySql.
Pelo que eu li o Node se da bem com diversos bancos de dados, mas parece que se criou uma cultura de usar ele com o MongoDB (me corrijam se eu estiver errado)

**Comment by raisiqueira** [comment]
Aqui estou usando NodeJS em conjunto com Express, Passport e Mongo(Mongoose)

**Comment by kivervinicius** [comment]
Trabalho atualmente com SqlServer e Couchbase, funciona tranquilamente, cada qual tem um FW para auxiliar o banco. Mas ouvi falar esses dias deste cara aqui https://github.com/typeorm/typeorm

**Comment by suissa** [comment]
Eu uso NodeJs, normalmente, com Express (mas pode ser qquer outro de rotas), Mongoose Socket.io e Passport, isso já dá uma boa base para iniciar qualquer sistema "simples".

**Comment by rictorres** [comment]
eu ja usei node.js em vários apps/sistemas em stacks bem diferentes e posso dizer que tudo depende do seu objetivo hahah

mas, como a maioria das linguagens atualmente, node.js trabalha muito bem com:
- mysql (recomendo [objection.js](https://github.com/Vincit/objection.js/) como orm)
- postgres (recomendo [objection.js](https://github.com/Vincit/objection.js/) como orm)
- mongodb
- redis (com [ioredis](https://github.com/luin/ioredis))
- rethinkdb
- websockets (veja nota abaixo)
- sse
- webrtc
- etc
- ...
- e muito mais!!! ☺️ 

sobre os web frameworks, ja usei:
- express
- koa
- hapi
- restify

confesso que nos ultimos grandes projetos usei o koa. ah.. tem também alguns frameworks mais completos (very opinionated) como o [nodal](http://www.nodaljs.com/), talvez valha a pena dar uma olhada :)

nota sobre websockets: por favor use o [`primus`](https://github.com/primus/primus) como uma abstração e [`uws`](https://github.com/uNetworking/uWebSockets) ou [`ws`](https://github.com/websockets/ws) como transporte. veja as principais [vantagens de usar o `primus`](https://github.com/primus/primus#advantages) e um [benchmark do `uws`](https://github.com/uNetworking/uWebSockets#not-your-average-server)

acho que é isso 😎

**Comment by Cleysonlb** [comment]
@rictorres cara muito obrigado, sua resposta foi super completa e me deu um direcionamento do que procurar! Vlw mesmo!!

**Comment by rafaelveiga** [comment]
Oi @Cleysonlb, acredito que me encontro na mesma situação que você.

Possuo uma API em produção em Node.JS que conversa com um banco de dados em MySQL, mas ainda não evolui a ponto de usar ORM, etc. 

O que venho usando de packages interessantes:
- **express-busboy** - uma boa alternativa ao body-parser se você deseja dar suporte a upload de arquivos. Fiz a troca do body-parser para este package no meio do projeto e tive 0 retrabalho nos meus métodos, já que a API que ele fornece é bem similar ao body-parser.
- **[node-mysql](https://github.com/mysqljs/mysql)** - package que faz toda a interface com MySQL. Apesar de não ter uma API interessante p/ montar as queries, o seu uso é bem prático. 

Ainda assim, acho que tenho muito a evoluir e aprender, e meu projeto está sempre evoluindo.

**Comment by tcelestino** [comment]
- Express || Hapi
- Mongoose

Próximo desafio é fazer um worker usando redis ou rabbitmq, mas ainda não decidir, esse é o grande problema. 😄

**Comment by leonardosouza** [comment]
Eu estava dando uma olhada esses dias no `Feathersjs`. Me parece bem promissor:

https://docs.feathersjs.com/guides/about/features.html

😃

**Comment by Lucaalmeida** [comment]
Gente sou iniciante em js e no node js podem me indicar algum curso com projetos

**Comment by cristofersousa** [comment]
@Lucaalmeida  segue > 

* [Youtube](https://www.youtube.com/playlist?list=PLJ_KhUnlXUPtbtLwaxxUxHqvcNQndmI4B)
* [NodeBR - Erick Wendel](https://cursos.nodebr.org/p/node-js-para-iniciantes-nodebr)

**Comment by Lucaalmeida** [comment]
@cristofersousa  Obrigado meu amigo vlw

**Comment by kavalcante** [comment]
A nossa stack aqui é a seguinte:

- Express para rotas simples
- GraphQL com Apollo
- JWT
- Mongoose como ORM

**Comment by Lucaalmeida** [comment]
Cursos com projetos o que vocês indicam, está difícil para 2k na Rocketseat

**Comment by kazzkiq** [comment]
Pra quem gosta de TypeScript como eu, recomendo a stack:

- APIs/Server: [typescript-rest](https://github.com/thiagobustamante/typescript-rest)
- IoC: [typescript-ioc](https://github.com/thiagobustamante/typescript-ioc)
- Banco Relacional: [TypeORM](https://github.com/typeorm/typeorm)
- MongoDB: [Typegoose](https://github.com/szokodiakos/typegoose)

Vantagens:
- Velocidade absurda na criação de endpoints usando Decorators e herança;
- Gera automaticamente o Swagger da API toda pra entregar pra equipe front-end;
- Permite controle total sobre os tipos/contrato que a API espera/envia.

**Comment by felipefialho** [comment]
Recomendo também dar uma olhada em Serverless com NodeJs
